File: src/ml/test-gp-vs-sklearn.js
/**
 * Compare ds GaussianProcessRegressor with scikit-learn reference values.
 * Reference values from running test-gp-vs-sklearn.py
 */
import { GaussianProcessRegressor } from './estimators/GaussianProcessRegressor.js';
import { RBF } from './kernels/rbf.js';

const TOLERANCE = 1e-6;

function assertClose(actual, expected, name, tol = TOLERANCE) {
  const diff = Math.abs(actual - expected);
  const pass = diff < tol;
  console.log(`${pass ? '✓' : '✗'} ${name}: ${actual.toFixed(8)} (expected ${expected.toFixed(8)}, diff=${diff.toExponential(2)})`);
  return pass;
}

console.log("=== Test 1: RBF Kernel Values (compare to sklearn) ===");
// sklearn reference: RBF with length_scale=1.0
// K[0,0] = 1.0, K[0,1] = 0.6065306597126334, K[0,2] = 0.1353352832366127
const kernel1 = new RBF({ lengthScale: 1.0, amplitude: 1.0 });

// Compute K([0], [0]) - should be 1.0
const k00 = kernel1.compute([0], [0]);
assertClose(k00, 1.0, "K([0], [0])");

// Compute K([0], [1]) - should be exp(-0.5 * 1^2) = 0.6065...
const k01 = kernel1.compute([0], [1]);
assertClose(k01, 0.6065306597126334, "K([0], [1])");

// Compute K([0], [2]) - should be exp(-0.5 * 2^2) = 0.1353...
const k02 = kernel1.compute([0], [2]);
assertClose(k02, 0.1353352832366127, "K([0], [2])");

console.log("\n=== Test 2: RBF Kernel with length_scale=2.0 ===");
// sklearn: K[0,1] with length_scale=2.0 = 0.8824969025845955
const kernel2 = new RBF({ lengthScale: 2.0, amplitude: 1.0 });
const k01_ls2 = kernel2.compute([0], [1]);
assertClose(k01_ls2, 0.8824969025845955, "K([0], [1]) with ls=2.0");

console.log("\n=== Test 3: Full Kernel Matrix ===");
// Build kernel matrix manually
const X = [[0], [1], [2]];
const K = X.map(x1 => X.map(x2 => kernel1.compute(x1, x2)));
console.log("Kernel matrix K(X, X):");
K.forEach((row, i) => console.log(`  [${row.map(v => v.toFixed(8)).join(', ')}]`));

// Compare to sklearn values
const sklearnK = [
  [1.0, 0.60653066, 0.13533528],
  [0.60653066, 1.0, 0.60653066],
  [0.13533528, 0.60653066, 1.0]
];
let kernelMatrixMatch = true;
for (let i = 0; i < 3; i++) {
  for (let j = 0; j < 3; j++) {
    if (Math.abs(K[i][j] - sklearnK[i][j]) > 1e-6) {
      kernelMatrixMatch = false;
    }
  }
}
console.log(kernelMatrixMatch ? "✓ Kernel matrix matches sklearn" : "✗ Kernel matrix differs from sklearn");

console.log("\n=== Test 4: GP Fit and Predict ===");
// sklearn reference (with optimizer=None, fixed kernel):
// X_train = [[0], [1], [2], [3], [4]], y_train = [0, 1, 0, 1, 0]
// alpha = 0.1, predictions at training points: [0.1216, 0.7229, 0.2757, 0.7229, 0.1216]
const xTrain = [[0], [1], [2], [3], [4]];
const yTrain = [0, 1, 0, 1, 0];

const gp = new GaussianProcessRegressor({ 
  kernel: new RBF({ lengthScale: 1.0, amplitude: 1.0 }), 
  alpha: 0.1 
});
gp.fit(xTrain, yTrain);

const { mean: yPred } = gp.predict(xTrain, { returnStd: true });
console.log("Predictions:", yPred.map(v => v.toFixed(8)));

// sklearn predictions (with optimizer=None): [0.12164727, 0.72287914, 0.27567115, 0.72287914, 0.12164727]
assertClose(yPred[0], 0.12164727, "Prediction at x=0", 0.01);
assertClose(yPred[1], 0.72287914, "Prediction at x=1", 0.01);
assertClose(yPred[2], 0.27567115, "Prediction at x=2", 0.01);
assertClose(yPred[3], 0.72287914, "Prediction at x=3", 0.01);
assertClose(yPred[4], 0.12164727, "Prediction at x=4", 0.01);

console.log("\n=== Test 5: Interpolation at Training Points ===");
// With low noise (alpha=0.01), predictions should be very close to training values
const gpTight = new GaussianProcessRegressor({ 
  kernel: new RBF({ lengthScale: 1.0, amplitude: 1.0 }), 
  alpha: 0.01 
});
gpTight.fit(xTrain, yTrain);
const { mean: yPredTight } = gpTight.predict(xTrain, { returnStd: true });

console.log("With alpha=0.01:");
console.log("  Training y:", yTrain);
console.log("  Predicted:", yPredTight.map(v => v.toFixed(4)));
const maxError = Math.max(...yTrain.map((y, i) => Math.abs(y - yPredTight[i])));
console.log(`  Max error: ${maxError.toFixed(6)}`);
console.log(maxError < 0.1 ? "✓ Good interpolation" : "✗ Poor interpolation");

console.log("\n=== Test 6: Predict at New Points ===");
// This tests the GP's ability to interpolate between training points
const xNew = [[0.5], [1.5], [2.5], [3.5]];
const { mean: yNew } = gp.predict(xNew, { returnStd: true });
console.log("Predictions at [0.5, 1.5, 2.5, 3.5]:", yNew.map(v => v.toFixed(4)));

// The pattern is [0, 1, 0, 1, 0], so midpoints should be around 0.5
// Actually with RBF kernel, the interpolation depends on the kernel bandwidth
console.log("Midpoint predictions should show smooth interpolation between training points");

console.log("\n=== Test 7: Posterior Sampling ===");
// Test that sampling from posterior gives values near predictions
const samples = gp.sample(xTrain, 3, 42);
console.log("Posterior samples (3 samples at training points):");
for (let i = 0; i < 3; i++) {
  console.log(`  Sample ${i + 1}: [${samples[i].map(v => v.toFixed(4)).join(', ')}]`);
}

// Samples should have mean close to predictions
const sampleMeans = xTrain.map((_, j) => {
  return samples.reduce((sum, s) => sum + s[j], 0) / 3;
});
console.log(`Sample means: [${sampleMeans.map(v => v.toFixed(4)).join(', ')}]`);
console.log(`Predictions:  [${yPred.map(v => v.toFixed(4)).join(', ')}]`);

console.log("\n=== Summary ===");
console.log("Key comparisons with scikit-learn:");
console.log(`  RBF kernel formula: k(x1, x2) = amplitude * exp(-0.5 * ||x1-x2||^2 / lengthScale^2)`);
console.log(`  sklearn uses:       k(x1, x2) = exp(-0.5 * ||x1-x2||^2 / length_scale^2)`);
console.log(`  → Both implementations should match when amplitude=1.0`);

File: src/ml/test-gp-vs-sklearn.py
"""
Compare ds GaussianProcessRegressor with scikit-learn's implementation.
Run this to get reference values for comparison.

IMPORTANT: Use optimizer=None to prevent sklearn from optimizing the kernel hyperparameters.
Otherwise sklearn changes length_scale and results won't match.
"""
import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF

np.random.seed(42)

# Test 1: RBF Kernel computation
print("=== Test 1: RBF Kernel Values ===")
kernel = RBF(length_scale=1.0)
kernel.length_scale_bounds = "fixed"  # Prevent optimization
X1 = np.array([[0], [1], [2]])
X2 = np.array([[0], [1], [2]])
K = kernel(X1, X2)
print("K(X, X) with length_scale=1.0:")
print(K)

# Test 2: Fit and predict
print("\n=== Test 2: Fit and Predict ===")
X_train = np.array([[0], [1], [2], [3], [4]]).reshape(-1, 1)
y_train = np.array([0, 1, 0, 1, 0])

# Use optimizer=None to disable hyperparameter optimization
gp = GaussianProcessRegressor(kernel=RBF(length_scale=1.0), alpha=0.1, optimizer=None, random_state=42)
gp.fit(X_train, y_train)

X_test = np.array([[0], [1], [2], [3], [4]])
y_pred, y_std = gp.predict(X_test, return_std=True)
print(f"Predictions: {y_pred}")
print(f"Std devs: {y_std}")

# Test 3: Predictions at training points should be close to training values
print("\n=== Test 3: Interpolation Accuracy ===")
print(f"Training y: {y_train}")
print(f"Predicted y: {y_pred}")
print(f"Max error: {np.max(np.abs(y_train - y_pred)):.6f}")

# Test 4: Kernel matrix properties
print("\n=== Test 4: Kernel Matrix Properties ===")
kernel2 = RBF(length_scale=2.0)
K2 = kernel2(X1, X2)
print("K(X, X) with length_scale=2.0:")
print(K2)

# Test 5: Variance at training points should be low
print("\n=== Test 5: Variance at Training Points ===")
gp_tight = GaussianProcessRegressor(kernel=RBF(length_scale=1.0), alpha=0.01, optimizer=None, random_state=42)
gp_tight.fit(X_train, y_train)
_, y_std_tight = gp_tight.predict(X_test, return_std=True)
print(f"Std devs with alpha=0.01: {y_std_tight}")

# Test 6: Predict at new points
print("\n=== Test 6: Predict at New Points ===")
X_new = np.array([[0.5], [1.5], [2.5], [3.5]])
y_new, y_std_new = gp.predict(X_new, return_std=True)
print(f"Predictions at [0.5, 1.5, 2.5, 3.5]: {y_new}")
print(f"Std devs: {y_std_new}")

# Test 7: Sample from posterior
print("\n=== Test 7: Sample from Posterior ===")
# Note: sklearn uses random_state during fit, but sample_y uses its own random state
samples = gp.sample_y(X_test, n_samples=3, random_state=42)
print(f"Sample shape: {samples.shape}")
print(f"Sample 1: {samples[:, 0]}")
print(f"Sample 2: {samples[:, 1]}")
print(f"Sample 3: {samples[:, 2]}")

print("\n=== Summary: Key Values for Comparison ===")
print("RBF(1.0) K[0,0]:", K[0, 0])
print("RBF(1.0) K[0,1]:", K[0, 1])
print("RBF(1.0) K[0,2]:", K[0, 2])
print("RBF(2.0) K[0,1]:", K2[0, 1])
print("Prediction at x=0:", y_pred[0])
print("Prediction at x=2:", y_pred[2])
